File: lib/dsp/stft.dart
import 'dart:ffi' as ffi;
import 'dart:typed_data';

import 'package:ffi/ffi.dart';

import '../bindings/stft_bindings.dart'
    show
        yl_stft_bins,
        yl_stft_cfg,
        yl_stft_create,
        yl_stft_destroy,
        yl_stft_flush_r2c,
        yl_stft_flush_spec,
        yl_stft_handle,
        yl_stft_hop,
        yl_stft_nfft,
        yl_stft_process_r2c,
        yl_stft_process_spec,
        yl_stft_push_r2c,
        yl_stft_push_spec,
        yl_stft_reset,
        yl_stft_win_len;
import '../utils/native_handle.dart';
import '../utils/types.dart';
import 'fft.dart' show SpecMode;

/// Window shape applied to each STFT frame before the FFT.
enum StftWindow {
  rect(0),
  hann(1),
  hamming(2),
  blackman(3);

  const StftWindow(this.value);
  final int value;
}

/// Complex STFT output stored as contiguous frame-major real and imaginary bins.
class STFTFrames {
  final int frames;
  final int bins;
  final Float32List real;
  final Float32List imag;

  const STFTFrames._(this.frames, this.bins, this.real, this.imag);

  /// Builds a complex STFT result from caller-owned real and imaginary buffers.
  factory STFTFrames.fromBuffers({
    required int frames,
    required int bins,
    required Float32List real,
    required Float32List imag,
  }) {
    final expectedLen = frames * bins;
    if (real.length != expectedLen || imag.length != expectedLen) {
      throw ArgumentError(
        'real/imag must both have length frames * bins ($expectedLen)',
      );
    }
    return STFTFrames._(frames, bins, real, imag);
  }

  /// Returns an empty complex STFT result with the given bin count.
  factory STFTFrames.empty(int bins) {
    return STFTFrames._(0, bins, Float32List(0), Float32List(0));
  }

  /// Whether this result contains zero frames.
  bool get isEmpty => frames == 0;

  /// Wraps the flattened real and imaginary buffers as a complex vector view.
  Complex32List asComplex() => Complex32List.wrap(real, imag);

  /// Returns the real bins for a single frame as a zero-copy sublist view.
  Float32List frameReal(int index) {
    if (index < 0 || index >= frames) {
      throw RangeError.index(index, real, 'index', null, frames);
    }
    final start = index * bins;
    return Float32List.sublistView(real, start, start + bins);
  }

  /// Returns the imaginary bins for a single frame as a zero-copy sublist view.
  Float32List frameImag(int index) {
    if (index < 0 || index >= frames) {
      throw RangeError.index(index, imag, 'index', null, frames);
    }
    final start = index * bins;
    return Float32List.sublistView(imag, start, start + bins);
  }
}

/// Scalar STFT output stored as contiguous frame-major bins.
class STFTSpecFrames {
  final int frames;
  final int bins;
  final Float32List values;

  const STFTSpecFrames._(this.frames, this.bins, this.values);

  /// Builds a scalar STFT result from a caller-owned frame-major buffer.
  factory STFTSpecFrames.fromBuffer({
    required int frames,
    required int bins,
    required Float32List values,
  }) {
    final expectedLen = frames * bins;
    if (values.length != expectedLen) {
      throw ArgumentError(
        'values must have length frames * bins ($expectedLen)',
      );
    }
    return STFTSpecFrames._(frames, bins, values);
  }

  /// Returns an empty scalar STFT result with the given bin count.
  factory STFTSpecFrames.empty(int bins) {
    return STFTSpecFrames._(0, bins, Float32List(0));
  }

  /// Whether this result contains zero frames.
  bool get isEmpty => frames == 0;

  /// Returns one scalar spectrum frame as a zero-copy sublist view.
  Float32List frame(int index) {
    if (index < 0 || index >= frames) {
      throw RangeError.index(index, values, 'index', null, frames);
    }
    final start = index * bins;
    return Float32List.sublistView(values, start, start + bins);
  }
}

final class _StftResource implements NativeResource {
  final yl_stft_handle handle;

  _StftResource(this.handle);

  /// Releases the native STFT context and its internal scratch buffers.
  @override
  void free() {
    yl_stft_destroy(handle);
  }
}

class STFT extends NativeHandle<_StftResource> {
  final int nFft;
  final int winLen;
  final int hop;
  final int bins;
  final StftWindow window;
  final bool meanSubtract;
  final bool center;

  STFT._({
    required this.nFft,
    required this.winLen,
    required this.hop,
    required this.bins,
    required this.window,
    required this.meanSubtract,
    required this.center,
    required yl_stft_handle handle,
  }) : super(_StftResource(handle));

  /// Creates a reusable native STFT context.
  ///
  /// The instance can be used in streaming mode via [push] and [flush], or in
  /// whole-buffer mode via [process]. The same context also exposes scalar
  /// spectrum variants through [pushSpec], [flushSpec], and [processSpec].
  factory STFT({
    required int nFft,
    int? winLen,
    required int hop,
    StftWindow window = StftWindow.hann,
    bool meanSubtract = false,
    bool center = false,
  }) {
    final actualWinLen = winLen ?? nFft;
    if (nFft <= 0) {
      throw ArgumentError.value(nFft, 'nFft', 'must be > 0');
    }
    if (actualWinLen <= 0 || actualWinLen > nFft) {
      throw ArgumentError.value(
        actualWinLen,
        'winLen',
        'must be > 0 and <= nFft',
      );
    }
    if (hop <= 0) {
      throw ArgumentError.value(hop, 'hop', 'must be > 0');
    }

    final cfg = calloc<yl_stft_cfg>();
    try {
      cfg.ref
        ..n_fft = nFft
        ..win_len = actualWinLen
        ..hop = hop
        ..win = window.value
        ..mean_subtract = meanSubtract ? 1 : 0
        ..center = center ? 1 : 0;

      final handle = yl_stft_create(cfg);
      if (handle == ffi.nullptr) {
        throw StateError(
          'yl_stft_create(nFft: $nFft, winLen: $actualWinLen, hop: $hop) returned nullptr',
        );
      }

      return STFT._(
        nFft: yl_stft_nfft(handle),
        winLen: yl_stft_win_len(handle),
        hop: yl_stft_hop(handle),
        bins: yl_stft_bins(handle),
        window: window,
        meanSubtract: meanSubtract,
        center: center,
        handle: handle,
      );
    } finally {
      calloc.free(cfg);
    }
  }

  /// Creates an STFT context intended for repeated chunked input.
  factory STFT.streaming({
    required int nFft,
    int? winLen,
    required int hop,
    StftWindow window = StftWindow.hann,
    bool meanSubtract = false,
    bool center = false,
  }) => STFT(
    nFft: nFft,
    winLen: winLen,
    hop: hop,
    window: window,
    meanSubtract: meanSubtract,
    center: center,
  );

  /// Creates an STFT context intended for whole-buffer processing.
  factory STFT.offline({
    required int nFft,
    int? winLen,
    required int hop,
    StftWindow window = StftWindow.hann,
    bool meanSubtract = false,
    bool center = false,
  }) => STFT(
    nFft: nFft,
    winLen: winLen,
    hop: hop,
    window: window,
    meanSubtract: meanSubtract,
    center: center,
  );

  /// Clears the streaming state so the next input starts a fresh analysis run.
  void reset() {
    yl_stft_reset(res.handle);
  }

  /// Maximum number of frames a flush can emit for the current overlap setup.
  int get maxFlushFrames {
    if (hop >= winLen) return 1;
    return (winLen + hop - 1) ~/ hop;
  }

  /// Conservative frame upper bound for one streaming push call.
  int maxPushFrames(int inputLength) {
    if (inputLength <= 0) return 0;
    return ((inputLength + hop - 1) ~/ hop) + 1;
  }

  /// Conservative frame upper bound for [process], including the final flush.
  int maxProcessFrames(int signalLength) {
    if (signalLength <= 0) return 0;
    return maxPushFrames(signalLength) + maxFlushFrames;
  }

  /// Pushes one input chunk and returns any complex STFT frames produced.
  STFTFrames push(Float32List input) {
    if (input.isEmpty) return STFTFrames.empty(bins);
    return _runInputComplexOp(
      input,
      framesCap: maxPushFrames(input.length),
      op: (inputPtr, outRe, outIm, framesCap) {
        return yl_stft_push_r2c(
          res.handle,
          inputPtr,
          input.length,
          outRe,
          outIm,
          framesCap,
        );
      },
    );
  }

  /// Flushes the tail of a streaming run and returns the remaining complex frames.
  STFTFrames flush() {
    return _runOutputOnlyComplexOp(
      framesCap: maxFlushFrames,
      op: (outRe, outIm, framesCap) {
        return yl_stft_flush_r2c(res.handle, outRe, outIm, framesCap);
      },
    );
  }

  /// Processes a whole signal buffer and returns all complex STFT frames.
  STFTFrames process(Float32List signal) {
    if (signal.isEmpty) {
      reset();
      return STFTFrames.empty(bins);
    }
    return _runInputComplexOp(
      signal,
      framesCap: maxProcessFrames(signal.length),
      op: (inputPtr, outRe, outIm, framesCap) {
        return yl_stft_process_r2c(
          res.handle,
          inputPtr,
          signal.length,
          outRe,
          outIm,
          framesCap,
        );
      },
    );
  }

  /// Pushes one input chunk and returns scalar spectrum frames directly.
  ///
  /// This path uses the native FFT spectrum routine, so it can emit magnitude,
  /// power, or dB bins without returning intermediate complex spectra.
  STFTSpecFrames pushSpec(
    Float32List input, {
    SpecMode mode = SpecMode.power,
    double dbFloor = -120.0,
  }) {
    if (input.isEmpty) return STFTSpecFrames.empty(bins);
    return _runInputSpecOp(
      input,
      framesCap: maxPushFrames(input.length),
      mode: mode,
      dbFloor: dbFloor,
      op: (inputPtr, outBins, framesCap, mode, dbFloor) {
        return yl_stft_push_spec(
          res.handle,
          inputPtr,
          input.length,
          outBins,
          framesCap,
          mode.value,
          dbFloor,
        );
      },
    );
  }

  /// Flushes the tail of a streaming run and returns scalar spectrum frames.
  STFTSpecFrames flushSpec({
    SpecMode mode = SpecMode.power,
    double dbFloor = -120.0,
  }) {
    return _runOutputOnlySpecOp(
      framesCap: maxFlushFrames,
      mode: mode,
      dbFloor: dbFloor,
      op: (outBins, framesCap, mode, dbFloor) {
        return yl_stft_flush_spec(
          res.handle,
          outBins,
          framesCap,
          mode.value,
          dbFloor,
        );
      },
    );
  }

  /// Processes a whole signal buffer and returns scalar spectrum frames directly.
  STFTSpecFrames processSpec(
    Float32List signal, {
    SpecMode mode = SpecMode.power,
    double dbFloor = -120.0,
  }) {
    if (signal.isEmpty) {
      reset();
      return STFTSpecFrames.empty(bins);
    }
    return _runInputSpecOp(
      signal,
      framesCap: maxProcessFrames(signal.length),
      mode: mode,
      dbFloor: dbFloor,
      op: (inputPtr, outBins, framesCap, mode, dbFloor) {
        return yl_stft_process_spec(
          res.handle,
          inputPtr,
          signal.length,
          outBins,
          framesCap,
          mode.value,
          dbFloor,
        );
      },
    );
  }

  /// Allocates native input/output buffers for one complex STFT call with input.
  STFTFrames _runInputComplexOp(
    Float32List input, {
    required int framesCap,
    required int Function(
      ffi.Pointer<ffi.Float> inputPtr,
      ffi.Pointer<ffi.Float> outRe,
      ffi.Pointer<ffi.Float> outIm,
      int framesCap,
    )
    op,
  }) {
    ffi.Pointer<ffi.Float> inputPtr = ffi.nullptr;
    ffi.Pointer<ffi.Float> outRePtr = ffi.nullptr;
    ffi.Pointer<ffi.Float> outImPtr = ffi.nullptr;

    try {
      inputPtr = calloc<ffi.Float>(input.length);
      inputPtr.asTypedList(input.length).setAll(0, input);

      final outputLen = framesCap * bins;
      outRePtr = calloc<ffi.Float>(outputLen);
      outImPtr = calloc<ffi.Float>(outputLen);

      final frames = op(inputPtr, outRePtr, outImPtr, framesCap);
      return _copyComplexFrames(frames, outRePtr, outImPtr);
    } finally {
      if (inputPtr != ffi.nullptr) calloc.free(inputPtr);
      if (outRePtr != ffi.nullptr) calloc.free(outRePtr);
      if (outImPtr != ffi.nullptr) calloc.free(outImPtr);
    }
  }

  /// Allocates native output buffers for one complex STFT call without input.
  STFTFrames _runOutputOnlyComplexOp({
    required int framesCap,
    required int Function(
      ffi.Pointer<ffi.Float> outRe,
      ffi.Pointer<ffi.Float> outIm,
      int framesCap,
    )
    op,
  }) {
    ffi.Pointer<ffi.Float> outRePtr = ffi.nullptr;
    ffi.Pointer<ffi.Float> outImPtr = ffi.nullptr;

    try {
      final outputLen = framesCap * bins;
      outRePtr = calloc<ffi.Float>(outputLen);
      outImPtr = calloc<ffi.Float>(outputLen);

      final frames = op(outRePtr, outImPtr, framesCap);
      return _copyComplexFrames(frames, outRePtr, outImPtr);
    } finally {
      if (outRePtr != ffi.nullptr) calloc.free(outRePtr);
      if (outImPtr != ffi.nullptr) calloc.free(outImPtr);
    }
  }

  /// Allocates native input/output buffers for one scalar-spectrum STFT call.
  STFTSpecFrames _runInputSpecOp(
    Float32List input, {
    required int framesCap,
    required SpecMode mode,
    required double dbFloor,
    required int Function(
      ffi.Pointer<ffi.Float> inputPtr,
      ffi.Pointer<ffi.Float> outBins,
      int framesCap,
      SpecMode mode,
      double dbFloor,
    )
    op,
  }) {
    ffi.Pointer<ffi.Float> inputPtr = ffi.nullptr;
    ffi.Pointer<ffi.Float> outBinsPtr = ffi.nullptr;

    try {
      inputPtr = calloc<ffi.Float>(input.length);
      inputPtr.asTypedList(input.length).setAll(0, input);

      final outputLen = framesCap * bins;
      outBinsPtr = calloc<ffi.Float>(outputLen);

      final frames = op(inputPtr, outBinsPtr, framesCap, mode, dbFloor);
      return _copySpecFrames(frames, outBinsPtr);
    } finally {
      if (inputPtr != ffi.nullptr) calloc.free(inputPtr);
      if (outBinsPtr != ffi.nullptr) calloc.free(outBinsPtr);
    }
  }

  /// Allocates native output buffers for one scalar-spectrum flush call.
  STFTSpecFrames _runOutputOnlySpecOp({
    required int framesCap,
    required SpecMode mode,
    required double dbFloor,
    required int Function(
      ffi.Pointer<ffi.Float> outBins,
      int framesCap,
      SpecMode mode,
      double dbFloor,
    )
    op,
  }) {
    ffi.Pointer<ffi.Float> outBinsPtr = ffi.nullptr;

    try {
      final outputLen = framesCap * bins;
      outBinsPtr = calloc<ffi.Float>(outputLen);

      final frames = op(outBinsPtr, framesCap, mode, dbFloor);
      return _copySpecFrames(frames, outBinsPtr);
    } finally {
      if (outBinsPtr != ffi.nullptr) calloc.free(outBinsPtr);
    }
  }

  /// Copies complex frame-major output from native memory into Dart-owned lists.
  STFTFrames _copyComplexFrames(
    int frames,
    ffi.Pointer<ffi.Float> outRePtr,
    ffi.Pointer<ffi.Float> outImPtr,
  ) {
    if (frames <= 0) return STFTFrames.empty(bins);
    final valuesLen = frames * bins;
    return STFTFrames.fromBuffers(
      frames: frames,
      bins: bins,
      real: Float32List.fromList(outRePtr.asTypedList(valuesLen)),
      imag: Float32List.fromList(outImPtr.asTypedList(valuesLen)),
    );
  }

  /// Copies scalar frame-major output from native memory into a Dart-owned list.
  STFTSpecFrames _copySpecFrames(
    int frames,
    ffi.Pointer<ffi.Float> outBinsPtr,
  ) {
    if (frames <= 0) return STFTSpecFrames.empty(bins);
    final valuesLen = frames * bins;
    return STFTSpecFrames.fromBuffer(
      frames: frames,
      bins: bins,
      values: Float32List.fromList(outBinsPtr.asTypedList(valuesLen)),
    );
  }
}

File: src/dsp/stft.c
#include "dsp/stft.h"
#include <stdlib.h>
#include <string.h>

struct yl_stft_ctx {
  yl_fft_plan* fft;

  uint32_t n_fft;
  uint32_t win_len;
  uint32_t hop;
  uint32_t bins;

  int mean_subtract;
  int center;

  yl_window_type win_type;
  float* window;       // win_len

  // Streaming buffer of size win_len
  float* ring;         // win_len
  uint32_t ring_fill;  // 0..win_len

  // Scratch FFT input length n_fft (could later come from backend getter)
  float* time;         // n_fft
};

static uint32_t yl_stft_center_pad(const yl_stft_ctx* s) {
  if (!s || !s->center) return 0u;
  return s->win_len / 2u;
}

static void yl_stft_prime_ring(yl_stft_ctx* s) {
  if (!s) return;
  s->ring_fill = yl_stft_center_pad(s);
  if (s->ring_fill > 0u) {
    memset(s->ring, 0, (size_t)s->ring_fill * sizeof(float));
  }
}

yl_stft_ctx* yl_stft_create(const yl_stft_cfg* cfg) {
  if (!cfg) return NULL;
  if (cfg->n_fft == 0 || cfg->win_len == 0 || cfg->hop == 0) return NULL;
  if (cfg->win_len > cfg->n_fft) return NULL;

  yl_stft_ctx* s = (yl_stft_ctx*)calloc(1, sizeof(*s));
  if (!s) return NULL;

  s->n_fft = cfg->n_fft;
  s->win_len = cfg->win_len;
  s->hop = cfg->hop;
  s->bins = (cfg->n_fft / 2u) + 1u;
  s->mean_subtract = cfg->mean_subtract ? 1 : 0;
  s->center = cfg->center ? 1 : 0;
  s->win_type = cfg->win;

  s->fft = yl_fft_create_r2c((size_t)cfg->n_fft);
  if (!s->fft) goto fail;

  s->window = (float*)malloc((size_t)s->win_len * sizeof(float));
  if (!s->window) goto fail;
  yl_window_generate(s->window, (int)s->win_len, s->win_type);

  s->ring = (float*)malloc((size_t)s->win_len * sizeof(float));
  if (!s->ring) goto fail;
  yl_stft_prime_ring(s);

  s->time = (float*)malloc((size_t)s->n_fft * sizeof(float));
  if (!s->time) goto fail;

  return s;

fail:
  yl_stft_destroy(s);
  return NULL;
}

void yl_stft_destroy(yl_stft_ctx* s) {
  if (!s) return;
  if (s->fft) yl_fft_destroy(s->fft);
  free(s->window);
  free(s->ring);
  free(s->time);
  free(s);
}

void yl_stft_reset(yl_stft_ctx* s) {
  if (!s) return;
  yl_stft_prime_ring(s);
  memset(s->time, 0, (size_t)s->n_fft * sizeof(float));
}

uint32_t yl_stft_bins(const yl_stft_ctx* s) {
  return s ? s->bins : 0u;
}

uint32_t yl_stft_nfft(const yl_stft_ctx* s) {
  return s ? s->n_fft : 0u;
}

uint32_t yl_stft_hop(const yl_stft_ctx* s) {
  return s ? s->hop : 0u;
}

uint32_t yl_stft_win_len(const yl_stft_ctx* s) {
  return s ? s->win_len : 0u;
}

static void yl_stft_prepare_frame(yl_stft_ctx* s, const float* frame) {
  // Optional mean subtraction
  float mean = 0.0f;
  if (s->mean_subtract) {
    for (uint32_t i = 0; i < s->win_len; ++i) mean += frame[i];
    mean /= (float)s->win_len;
  }

  // Window into time buffer
  for (uint32_t i = 0; i < s->win_len; ++i) {
    float v = frame[i] - mean;
    s->time[i] = v * s->window[i];
  }

  // Zero pad
  if (s->win_len < s->n_fft) {
    memset(s->time + s->win_len, 0, (size_t)(s->n_fft - s->win_len) * sizeof(float));
  }
}

static void yl_stft_advance_ring_full(yl_stft_ctx* s) {
  if (s->hop >= s->win_len) {
    s->ring_fill = 0;
  } else {
    const uint32_t keep = s->win_len - s->hop;
    memmove(s->ring, s->ring + s->hop, (size_t)keep * sizeof(float));
    s->ring_fill = keep;
  }
}

size_t yl_stft_push_r2c(
  yl_stft_ctx* s,
  const float* input,
  size_t input_len,
  float* out_re,
  float* out_im,
  size_t out_frames_cap
) {
  if (!s || (!input && input_len)) return 0;
  if (out_frames_cap && (!out_re || !out_im)) return 0;

  size_t frames = 0;
  size_t idx = 0;

  while (idx < input_len) {
    const uint32_t need = s->win_len - s->ring_fill;
    const size_t take = (input_len - idx < (size_t)need) ? (input_len - idx) : (size_t)need;

    memcpy(s->ring + s->ring_fill, input + idx, take * sizeof(float));
    s->ring_fill += (uint32_t)take;
    idx += take;

    if (s->ring_fill < s->win_len) break;
    if (frames >= out_frames_cap) break;

    // Prepare + FFT
    yl_stft_prepare_frame(s, s->ring);
    yl_fft_execute_r2c(s->fft, s->time,
                      out_re + frames * s->bins,
                      out_im + frames * s->bins);
    frames++;

    // Advance by hop
    yl_stft_advance_ring_full(s);
  }

  return frames;
}

size_t yl_stft_push_spec(
  yl_stft_ctx* s,
  const float* input,
  size_t input_len,
  float* out_bins,
  size_t out_frames_cap,
  yl_spec_mode mode,
  float db_floor
) {
  if (!s || (!input && input_len)) return 0;
  if (out_frames_cap && !out_bins) return 0;

  size_t frames = 0;
  size_t idx = 0;

  while (idx < input_len) {
    const uint32_t need = s->win_len - s->ring_fill;
    const size_t take = (input_len - idx < (size_t)need) ? (input_len - idx) : (size_t)need;

    memcpy(s->ring + s->ring_fill, input + idx, take * sizeof(float));
    s->ring_fill += (uint32_t)take;
    idx += take;

    if (s->ring_fill < s->win_len) break;
    if (frames >= out_frames_cap) break;

    yl_stft_prepare_frame(s, s->ring);
    yl_fft_execute_r2c_spec(
      s->fft,
      s->time,
      out_bins + frames * s->bins,
      mode,
      db_floor
    );
    frames++;

    yl_stft_advance_ring_full(s);
  }

  return frames;
}

size_t yl_stft_flush_r2c(
  yl_stft_ctx* s,
  float* out_re,
  float* out_im,
  size_t out_frames_cap
) {
  if (!s) return 0;
  if (out_frames_cap && (!out_re || !out_im)) return 0;

  size_t frames = 0;

  while (s->ring_fill > 0 && frames < out_frames_cap) {
    const uint32_t valid = s->ring_fill;

    // Pad remaining samples with zeros to complete a frame
    if (valid < s->win_len) {
      memset(s->ring + valid, 0,
             (size_t)(s->win_len - valid) * sizeof(float));
      s->ring_fill = s->win_len;
    }

    yl_stft_prepare_frame(s, s->ring);
    yl_fft_execute_r2c(s->fft, s->time,
                      out_re + frames * s->bins,
                      out_im + frames * s->bins);
    frames++;

    // Advance by hop
    if (s->hop >= valid) {
      s->ring_fill = 0;
    } else {
      const uint32_t keep = valid - s->hop;
      memmove(s->ring, s->ring + s->hop, (size_t)keep * sizeof(float));
      s->ring_fill = keep;
    }
  }

  return frames;
}

size_t yl_stft_flush_spec(
  yl_stft_ctx* s,
  float* out_bins,
  size_t out_frames_cap,
  yl_spec_mode mode,
  float db_floor
) {
  if (!s) return 0;
  if (out_frames_cap && !out_bins) return 0;

  size_t frames = 0;

  while (s->ring_fill > 0 && frames < out_frames_cap) {
    const uint32_t valid = s->ring_fill;

    if (valid < s->win_len) {
      memset(s->ring + valid, 0,
             (size_t)(s->win_len - valid) * sizeof(float));
      s->ring_fill = s->win_len;
    }

    yl_stft_prepare_frame(s, s->ring);
    yl_fft_execute_r2c_spec(
      s->fft,
      s->time,
      out_bins + frames * s->bins,
      mode,
      db_floor
    );
    frames++;

    if (s->hop >= valid) {
      s->ring_fill = 0;
    } else {
      const uint32_t keep = valid - s->hop;
      memmove(s->ring, s->ring + s->hop, (size_t)keep * sizeof(float));
      s->ring_fill = keep;
    }
  }

  return frames;
}

size_t yl_stft_process_r2c(
  yl_stft_ctx* s,
  const float* signal,
  size_t signal_len,
  float* out_re,
  float* out_im,
  size_t out_frames_cap
) {
  if (!s || (!signal && signal_len)) return 0;

  yl_stft_reset(s);

  size_t frames = 0;
  frames += yl_stft_push_r2c(s, signal, signal_len,
                             out_re, out_im,
                             out_frames_cap);

  if (frames < out_frames_cap) {
    frames += yl_stft_flush_r2c(s,
                                out_re + frames * s->bins,
                                out_im + frames * s->bins,
                                out_frames_cap - frames);
  }
  return frames;
}

size_t yl_stft_process_spec(
  yl_stft_ctx* s,
  const float* signal,
  size_t signal_len,
  float* out_bins,
  size_t out_frames_cap,
  yl_spec_mode mode,
  float db_floor
) {
  if (!s || (!signal && signal_len)) return 0;

  yl_stft_reset(s);

  size_t frames = 0;
  frames += yl_stft_push_spec(
    s,
    signal,
    signal_len,
    out_bins,
    out_frames_cap,
    mode,
    db_floor
  );

  if (frames < out_frames_cap) {
    frames += yl_stft_flush_spec(
      s,
      out_bins + frames * s->bins,
      out_frames_cap - frames,
      mode,
      db_floor
    );
  }
  return frames;
}
